File: backend/app/routers/members.py
"""HTTP routes for scoped member panel and care-work operations."""

from typing import Annotated, Literal

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_session
from app.core.security import get_current_coordinator
from app.models.entities import Coordinator
from app.schemas.members import (
    AssignmentRequest,
    CareGapResponse,
    CarePlanGoalResponse,
    CloseGapRequest,
    GoalStatusRequest,
    MemberDetailResponse,
    MemberPanelResponse,
    OutreachCreateRequest,
    OutreachResponse,
)
from app.services.member_service import assign_member, close_gap, create_outreach, get_member_detail, list_members, update_goal_status

router = APIRouter(prefix="/api/v1/members", tags=["members"])

SessionDependency = Annotated[AsyncSession, Depends(get_session)]
ActorDependency = Annotated[Coordinator, Depends(get_current_coordinator)]


@router.get("", response_model=MemberPanelResponse, summary="List an authorized member panel")
async def get_members(
    session: SessionDependency,
    actor: ActorDependency,
    query: Annotated[str | None, Query(max_length=120)] = None,
    risk_level: Annotated[str | None, Query(max_length=24)] = None,
    plan: Annotated[str | None, Query(max_length=120)] = None,
    coordinator_id: int | None = None,
    sort_by: Literal["last_name", "risk_level", "plan", "date_of_birth"] = "last_name",
    sort_order: Literal["asc", "desc"] = "asc",
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> MemberPanelResponse:
    """Return a role-scoped page of members with filters and ordering."""

    return await list_members(session, actor, query, risk_level, plan, coordinator_id, sort_by, sort_order, limit, offset)


@router.get("/{member_id}", response_model=MemberDetailResponse, summary="View an authorized member detail")
async def get_member(member_id: int, session: SessionDependency, actor: ActorDependency) -> MemberDetailResponse:
    """Return one masked member profile and record the view audit."""

    return await get_member_detail(session, actor, member_id)


@router.patch("/{member_id}/assignment", response_model=MemberDetailResponse, summary="Assign a member")
async def patch_assignment(
    member_id: int,
    request: AssignmentRequest,
    session: SessionDependency,
    actor: ActorDependency,
) -> MemberDetailResponse:
    """Assign a member when the signed-in supervisor is authorized."""

    return await assign_member(session, actor, member_id, request)


@router.post("/{member_id}/gaps/{gap_id}/close", response_model=CareGapResponse, summary="Close an open care gap")
async def post_close_gap(
    member_id: int,
    gap_id: int,
    request: CloseGapRequest,
    session: SessionDependency,
    actor: ActorDependency,
) -> CareGapResponse:
    """Close an open member care gap with a required reason."""

    return await close_gap(session, actor, member_id, gap_id, request.reason)


@router.post("/{member_id}/outreach", response_model=OutreachResponse, status_code=status.HTTP_201_CREATED, summary="Log outreach")
async def post_outreach(
    member_id: int,
    request: OutreachCreateRequest,
    session: SessionDependency,
    actor: ActorDependency,
) -> OutreachResponse:
    """Create a validated member outreach record."""

    return await create_outreach(session, actor, member_id, request)


@router.patch("/{member_id}/care-plan/{goal_id}", response_model=CarePlanGoalResponse, summary="Update a care-plan goal")
async def patch_care_plan_goal(
    member_id: int,
    goal_id: int,
    request: GoalStatusRequest,
    session: SessionDependency,
    actor: ActorDependency,
) -> CarePlanGoalResponse:
    """Update an authorized member goal to an allowed status."""

    return await update_goal_status(session, actor, member_id, goal_id, request)

File: frontend/src/api/members.ts
import type { CareGap, CarePlanGoal, MemberDetail, MemberPanel, Outreach, OutreachCreatePayload } from '../types';

/** Build a same-origin authenticated request and surface API errors. */
async function memberRequest<T>(path: string, token: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${import.meta.env.VITE_API_URL ?? ''}${path}`, {
    ...init,
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
      ...init?.headers,
    },
  });
  if (!response.ok) {
    const detail = await response.json().catch(() => ({ detail: 'The request could not be completed.' }));
    throw new Error(detail.detail ?? 'The request could not be completed.');
  }
  return response.json() as Promise<T>;
}

/** Fetch a bounded and scoped member panel using the current URL controls. */
export function getMembers(token: string, params: URLSearchParams): Promise<MemberPanel> {
  return memberRequest<MemberPanel>(`/api/v1/members?${params.toString()}`, token);
}

/** Fetch a masked detail profile for an authorized member. */
export function getMember(token: string, memberId: string): Promise<MemberDetail> {
  return memberRequest<MemberDetail>(`/api/v1/members/${memberId}`, token);
}

/** Close a member gap with required documentation. */
export function closeGap(token: string, memberId: number, gapId: number, reason: string): Promise<CareGap> {
  return memberRequest<CareGap>(`/api/v1/members/${memberId}/gaps/${gapId}/close`, token, {
    method: 'POST',
    body: JSON.stringify({ reason }),
  });
}

/** Record an outreach event for a member. */
export function createOutreach(token: string, memberId: number, payload: OutreachCreatePayload): Promise<Outreach> {
  return memberRequest<Outreach>(`/api/v1/members/${memberId}/outreach`, token, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

/** Assign or unassign a member when the signed-in user is a supervisor. */
export function updateMemberAssignment(token: string, memberId: number, coordinatorId: number | null): Promise<MemberDetail> {
  return memberRequest<MemberDetail>(`/api/v1/members/${memberId}/assignment`, token, {
    method: 'PATCH',
    body: JSON.stringify({ coordinator_id: coordinatorId }),
  });
}

/** Update an allowed care-plan goal status. */
export function updateGoalStatus(token: string, memberId: number, goalId: number, status: CarePlanGoal['status']): Promise<CarePlanGoal> {
  return memberRequest<CarePlanGoal>(`/api/v1/members/${memberId}/care-plan/${goalId}`, token, {
    method: 'PATCH',
    body: JSON.stringify({ status }),
  });
}
